theano/sandbox/gpuarray/nerv.py

```diff
@@ -0,0 +1,203 @@
+
+
+def ensure_float(val, name):
+    if not isinstance(val, Variable):
+        val = constant(val)
+    if hasattr(val, 'ndim') and val.ndim == 0:
+        val = as_scalar(val)
+    if not isinstance(val.type, theano.scalar.Scalar):
+        raise TypeError("%s: expected a scalar value" % (name,))
+    if not val.type.dtype == 'float32':
+        raise TypeError("%s: type is not float32" % (name,))
+    return val
+
+
+class Gemm16(COp):
+    __props__ = ('relu', 'inplace')
+    _f16_ok = True
+    KERN_NAMES = ('nn_128x128', 'nn_128x64', 'nn_128x32',
+                  'nn_vec_128x128', 'nn_vec_128x64', 'nn_vec_128x32',
+                  'tn_128x128', 'tn_128x64', 'tn_128x32',
+                  'tn_vec_128x128', 'tn_vec_128x64', 'tn_vec_128x32',
+                  'tn_vec_128x16', 'nt_128x128', 'nt_vec_128x128')
+
+    def __init__(self, relu=False, inplace=False):
+        COp.__init__(self, ["gemm16.c"], "gemm16")
+        self.relu = relu
+        # relu = True will require more work in optimizations.
+        assert self.relu is False
+        self.inplace = inplace
+        if self.inplace:
+            self.destroy_map = {0: [0]}
+
+    def make_node(self, C, alpha, A, B, beta):
+        if GPUTensor is None:
+            raise RuntimeError("Can't use Gemm16: nervanagpu not found")
+
+        A = as_gpuarray_variable(A)
+        B = as_gpuarray_variable(B)
+        C = as_gpuarray_variable(C)
+
+        alpha = ensure_float(alpha, 'alpha')
+        beta = ensure_float(beta, 'beta')
+
+        assert C.dtype == A.dtype == B.dtype == 'float16'
+
+        return Apply(self, [C, alpha, A, B, beta], [C.type()])
+
+    def perform(self, node, inputs, outputs):
+        ensure_pycuda_context()
+        C, alpha, A, B, beta = inputs
+        # The nervana code does not support the case where both inputs
+        # are trans, so we need to copy one if them if that is the
+        # case. We copy the smaller one.
+        if A.flags.f_contiguous and B.flags.f_contiguous:
+            if A.size < B.size:
+                A = A.copy()
+            else:
+                B = B.copy()
+        inplace = self.inplace
+        if inplace and not C.flags.c_contiguous:
+            inplace = False
+        if not inplace:
+            C = C.copy()
+        At = to_gputensor(A)
+        Bt = to_gputensor(B)
+        Ct = to_gputensor(C)
+        nerv.dot(At, Bt, Ct, alpha=alpha, beta=beta, relu=False)
+        outputs[0][0] = C
+
+    def c_headers(self):
+        return ['gpuarray/types.h', 'numpy_compat.h', 'gpuarray_helper.h',
+                'string.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__)]
+
+    def get_op_params(self):
+        return [('GEMM16_INPLACE', '1' if self.inplace else '0')]
+
+    @staticmethod
+    def cubin_to_code(name):
+        fname = 'hgemm_{0}.cubin'.format(name)
+        with open(os.path.join(nerv.cubin_path, fname)) as f:
```

Comment:
This fail with my old version of nervanagpu and the new one. The current filename you try to load:

nervanagpu/nervanagpu/kernels/cu/hgemm_nn_128x128.cubin

the current real filename:

nervanagpu/nervanagpu/kernels/cu/hgemm_nn_128x128.cu